programs/stake/src/stake_state.rs

```diff
@@ -1158,6 +1144,135 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
     }
 }
 
+#[derive(Clone, Debug, PartialEq)]
+enum MergeKind {
+    Inactive(Meta, u64),
+    ActivationEpoch(Meta, Stake),
+    FullyActive(Meta, Stake),
+}
+
+impl MergeKind {
+    fn meta(&self) -> &Meta {
+        match self {
+            Self::Inactive(meta, _) => meta,
+            Self::ActivationEpoch(meta, _) => meta,
+            Self::FullyActive(meta, _) => meta,
+        }
+    }
+
+    fn stake(&self) -> Option<&Stake> {
+        match self {
+            Self::Inactive(_, _) => None,
+            Self::ActivationEpoch(_, stake) => Some(stake),
+            Self::FullyActive(_, stake) => Some(stake),
+        }
+    }
+
+    fn get_if_mergeable(
+        stake_keyed_account: &KeyedAccount,
+        clock: &Clock,
+        stake_history: &StakeHistory,
+    ) -> Result<Self, InstructionError> {
+        match stake_keyed_account.state()? {
+            StakeState::Stake(meta, stake) => {
+                // stake must not be in a transient state. Transient here meaning
+                // activating or deactivating with non-zero effective stake.
+                match stake.delegation.stake_activating_and_deactivating(
+                    clock.epoch,
+                    Some(stake_history),
+                    true,
+                ) {
+                    (0, 0, 0) => Ok(Self::Inactive(meta, stake_keyed_account.lamports()?)),
+                    (0, _, _) => Ok(Self::ActivationEpoch(meta, stake)),
+                    (_, 0, 0) => Ok(Self::FullyActive(meta, stake)),
+                    _ => Err(StakeError::MergeActivatedStake.into()),
```

Comment:
Oh, speaking of `StakeErorr::MergeActivatedStake`, can we fix up this error message/display to be correct? I think editing the error variant should be okay as long as the index remains the same, right?

Reply:
Oh!  Good call!  I've been meaning to rename this to `MergeTransientStake`.  Will toss this in with Ryo's nits